imxrt-hal/src/adc.rs

```diff
@@ -223,6 +223,205 @@ where
+        AdcSource<ADCx, P>,
+        u16,
+        crate::dma::Linear<u16>,
+        crate::dma::Linear<u16>,
+    >,
+    unhandled_buffer_count: core::sync::atomic::AtomicUsize,
+    off_buffer: crate::dma::Linear<u16>,
+    _pin: PhantomData<P>,
+}
+
+impl<ADCx, P> StreamingAdc<ADCx, P>
+where
+    ADCx: adc::ADC + AdcDmaSource,
+    P: Pin<ADCx>,
+{
+    /// Create a new StreamingAdc object. The returned object will not begin
+    /// generating samples until start() is called.
+    pub fn new(
+        adc: ADC<ADCx>,
+        pin: AnalogInput<ADCx, P>,
+        channel: crate::dma::Channel,
+        buffer0: &'static crate::dma::Buffer<[u16; 256]>,
+        buffer1: &'static crate::dma::Buffer<[u16; 256]>,
+    ) -> Self {
+        let mut off_buffer = crate::dma::Linear::new(&buffer0).unwrap();
+        off_buffer.set_transfer_len(256);
+        let mut peripheral = crate::dma::receive_u16(AdcSource::new(adc, pin), channel);
+
+        Self {
+            buffer0,
+            buffer1,
+            peripheral,
+            unhandled_buffer_count: core::sync::atomic::AtomicUsize::new(0),
+            off_buffer,
+            _pin: PhantomData,
+        }
+    }
+
+    /// Begin transferring samples from the ADC. Once samples are flowing, the
+    /// user must call poll() frequently enough to keep up with the generated
+    /// samples, or else the interrupt handler will panic.
+    pub fn start(&mut self) {
+        let mut buffer = crate::dma::Linear::new(&self.buffer1).unwrap();
+        buffer.set_transfer_len(256);
+        self.peripheral.start_receive(buffer).unwrap();
+    }
+
+    /// Process any received samples from the ADC. If no samples have been
+    /// received since the last poll() call, cb() will not be called.
+    pub fn poll<F: FnOnce(&[u16])>(&mut self, cb: F) {
+        if self
+            .unhandled_buffer_count
+            .load(core::sync::atomic::Ordering::SeqCst)
+            > 0
+        {
+            cb(self.off_buffer.as_elements());
+            self.unhandled_buffer_count
+                .store(0, core::sync::atomic::Ordering::SeqCst);
+        }
+    }
+
+    /// Called from an interrupt handler. The user must register the appropriate
+    /// DMA handler for their channel and call this method from it.
+    ///
+    /// If poll() is not called frequently enough to keep up with the data, this
+    /// handler will panic.
+    pub fn handle_interrupt(&mut self) {
+        if self.peripheral.is_receive_interrupt() {
+            self.peripheral.receive_clear_interrupt();
+        }
+
+        let mut rx_buffer = self.peripheral.receive_complete().unwrap();
+        core::mem::swap(&mut self.off_buffer, &mut rx_buffer);
+        self.peripheral.start_receive(rx_buffer).unwrap();
+        if self
+            .unhandled_buffer_count
+            .swap(1, core::sync::atomic::Ordering::SeqCst)
+            > 0
+        {
+            panic!("User is not processing ADC data fast enough!");
+        }
```

Comment:
Should we check [`is_receive_complete()`](https://github.com/imxrt-rs/imxrt-hal/blob/8f457edaad37cfd294e6eaeae42605f3214a1b2e/imxrt-hal/src/dma/peripheral.rs#L209-L212) before completing the receive operation, and re-starting the transfer? Otherwise, a `handle_interrupt` call that's invoked before the transfer completes will panic at `receive_complete().unwrap()`.

(Concern is that we could call `handle_interrupt()` prematurely in an ISR. Consider the interrupt in the `StreamingAdc` example, and pretend that we're also using DMA channel 24 for something else. As written, we would invoke `handle_interrupt()` when channel 24 completes, which isn't necessarily when channel 8 completes.)

Reply:
> Should we check is_receive_complete() before completing the receive operation, and re-starting the transfer?

Yes! Good catch.